File: regions/system/library/go/migration/analyzer_test.go
package migration

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

// DROP TABLEのSQLからテーブル削除の破壊的変更が検出されることを確認する。
func TestDetectDropTable(t *testing.T) {
	sql := "DROP TABLE users;"
	changes := DetectBreakingChanges(sql)
	assert.Equal(t, 1, len(changes))
	assert.Equal(t, TableDropped, changes[0].Type)
	assert.Equal(t, "users", changes[0].Table)
}

// DROP COLUMNのSQLからカラム削除の破壊的変更が検出されることを確認する。
func TestDetectDropColumn(t *testing.T) {
	sql := "ALTER TABLE users DROP COLUMN email;"
	changes := DetectBreakingChanges(sql)
	assert.Equal(t, 1, len(changes))
	assert.Equal(t, ColumnDropped, changes[0].Type)
	assert.Equal(t, "users", changes[0].Table)
	assert.Equal(t, "email", changes[0].Column)
}

// SET NOT NULLのSQLからNOT NULL制約追加の破壊的変更が検出されることを確認する。
func TestDetectSetNotNull(t *testing.T) {
	sql := "ALTER TABLE users ALTER COLUMN email SET NOT NULL;"
	changes := DetectBreakingChanges(sql)
	assert.Equal(t, 1, len(changes))
	assert.Equal(t, NotNullAdded, changes[0].Type)
	assert.Equal(t, "users", changes[0].Table)
	assert.Equal(t, "email", changes[0].Column)
}

// SET DATA TYPEのSQLからカラム型変更の破壊的変更が検出されることを確認する。
func TestDetectTypeChange(t *testing.T) {
	sql := "ALTER TABLE users ALTER COLUMN age SET DATA TYPE BIGINT;"
	changes := DetectBreakingChanges(sql)
	assert.Equal(t, 1, len(changes))
	assert.Equal(t, ColumnTypeChanged, changes[0].Type)
	assert.Equal(t, "users", changes[0].Table)
	assert.Equal(t, "age", changes[0].Column)
	assert.Equal(t, "BIGINT", changes[0].To)
}

// RENAME COLUMNのSQLからカラム名変更の破壊的変更が検出されることを確認する。
func TestDetectRenameColumn(t *testing.T) {
	sql := "ALTER TABLE users RENAME COLUMN old_name TO new_name;"
	changes := DetectBreakingChanges(sql)
	assert.Equal(t, 1, len(changes))
	assert.Equal(t, ColumnRenamed, changes[0].Type)
	assert.Equal(t, "users", changes[0].Table)
	assert.Equal(t, "old_name", changes[0].From)
	assert.Equal(t, "new_name", changes[0].To)
}

// 非破壊的なSQLに対してDetectBreakingChangesが空のスライスを返すことを確認する。
func TestNoBreakingChanges(t *testing.T) {
	sql := "ALTER TABLE users ADD COLUMN email TEXT;"
	changes := DetectBreakingChanges(sql)
	assert.Empty(t, changes)
}

// BreakingChangeのStringメソッドが人間可読な表示文字列を返すことを確認する。
func TestDisplayFormatting(t *testing.T) {
	change := BreakingChange{
		Type:   ColumnDropped,
		Table:  "users",
		Column: "email",
	}
	assert.Equal(t, "Column users.email dropped", change.String())
}

// 不正なSQLを入力した場合にDetectBreakingChangesが空のスライスを返すことを確認する。
func TestInvalidSQLReturnsEmpty(t *testing.T) {
	changes := DetectBreakingChanges("NOT VALID SQL AT ALL !!!")
	assert.Empty(t, changes)
}

File: regions/system/library/typescript/migration/__tests__/analyzer.test.ts
import { describe, it, expect } from 'vitest';
import { detectBreakingChanges, formatBreakingChange } from '../src/analyzer.js';

describe('detectBreakingChanges', () => {
  it('should detect DROP TABLE', () => {
    const changes = detectBreakingChanges('DROP TABLE users;');
    expect(changes).toHaveLength(1);
    expect(changes[0]).toEqual({ type: 'tableDropped', table: 'users' });
  });

  it('should detect DROP COLUMN', () => {
    const changes = detectBreakingChanges(
      'ALTER TABLE users DROP COLUMN email;',
    );
    expect(changes).toHaveLength(1);
    expect(changes[0]).toEqual({
      type: 'columnDropped',
      table: 'users',
      column: 'email',
    });
  });

  it('should detect SET NOT NULL', () => {
    const changes = detectBreakingChanges(
      'ALTER TABLE users ALTER COLUMN email SET NOT NULL;',
    );
    expect(changes).toHaveLength(1);
    expect(changes[0]).toEqual({
      type: 'notNullAdded',
      table: 'users',
      column: 'email',
    });
  });

  it('should detect type change', () => {
    const changes = detectBreakingChanges(
      'ALTER TABLE users ALTER COLUMN age SET DATA TYPE BIGINT;',
    );
    expect(changes).toHaveLength(1);
    expect(changes[0]).toEqual({
      type: 'columnTypeChanged',
      table: 'users',
      column: 'age',
      from: 'unknown',
      to: 'BIGINT',
    });
  });

  it('should detect RENAME COLUMN', () => {
    const changes = detectBreakingChanges(
      'ALTER TABLE users RENAME COLUMN old_name TO new_name;',
    );
    expect(changes).toHaveLength(1);
    expect(changes[0]).toEqual({
      type: 'columnRenamed',
      table: 'users',
      from: 'old_name',
      to: 'new_name',
    });
  });

  it('should return no breaking changes for ADD COLUMN', () => {
    const changes = detectBreakingChanges(
      'ALTER TABLE users ADD COLUMN email TEXT;',
    );
    expect(changes).toHaveLength(0);
  });

  it('should return empty array for invalid SQL', () => {
    const changes = detectBreakingChanges('NOT VALID SQL AT ALL !!!');
    expect(changes).toHaveLength(0);
  });
});

describe('formatBreakingChange', () => {
  it('should format columnDropped', () => {
    expect(
      formatBreakingChange({
        type: 'columnDropped',
        table: 'users',
        column: 'email',
      }),
    ).toBe('Column users.email dropped');
  });

  it('should format columnTypeChanged', () => {
    expect(
      formatBreakingChange({
        type: 'columnTypeChanged',
        table: 'users',
        column: 'age',
        from: 'INT',
        to: 'BIGINT',
      }),
    ).toBe('Column users.age type changed from INT to BIGINT');
  });

  it('should format tableDropped', () => {
    expect(
      formatBreakingChange({ type: 'tableDropped', table: 'users' }),
    ).toBe('Table users dropped');
  });

  it('should format notNullAdded', () => {
    expect(
      formatBreakingChange({
        type: 'notNullAdded',
        table: 'users',
        column: 'email',
      }),
    ).toBe('NOT NULL added to users.email');
  });

  it('should format columnRenamed', () => {
    expect(
      formatBreakingChange({
        type: 'columnRenamed',
        table: 'users',
        from: 'old_name',
        to: 'new_name',
      }),
    ).toBe('Column users.old_name renamed to new_name');
  });
});
